programs/drift/src/validation/sig_verification.rs

```diff
@@ -100,72 +143,146 @@ pub fn verify_ed25519_msg<const N: usize>(
+
+    let expected_message_data_offset = expected_message_size_offset
+        .checked_add(MESSAGE_SIZE_LEN)
+        .ok_or(SignatureVerificationError::MessageOffsetOverflow)?;
+    if offsets.message_data_offset != expected_message_data_offset {
+        return Err(SignatureVerificationError::InvalidMessageOffset.into());
     }
 
-    Ok(())
-}
+    let expected_message_size: u16 = {
+        let start = usize::from(
+            expected_message_size_offset
+                .checked_sub(message_offset)
+                .unwrap(),
+        );
+        let end = usize::from(
+            expected_message_data_offset
+                .checked_sub(message_offset)
+                .unwrap(),
+        );
+        LE::read_u16(&msg[start..end])
+    };
+    if offsets.message_data_size != expected_message_size {
+        return Err(SignatureVerificationError::InvalidMessageDataSize.into());
+    }
+    if offsets.signature_instruction_index != current_ix_index
+        || offsets.public_key_instruction_index != current_ix_index
+        || offsets.message_instruction_index != current_ix_index
+    {
+        return Err(SignatureVerificationError::InvalidInstructionIndex.into());
+    }
 
-/// Extract pubkey from serialized Ed25519Program instruction data
-pub fn extract_ed25519_ix_pubkey(ix_data: &[u8]) -> Result<[u8; 32]> {
-    match ix_data[16..16 + 32].try_into() {
-        Ok(raw) => Ok(raw),
-        Err(_) => Err(ErrorCode::SigVerificationFailed.into()),
+    let public_key = {
+        let start = usize::from(
+            expected_public_key_offset
+                .checked_sub(message_offset)
+                .unwrap(),
+        );
+        let end = start
+            .checked_add(anchor_lang::solana_program::pubkey::PUBKEY_BYTES)
+            .ok_or(SignatureVerificationError::MessageOffsetOverflow)?;
+        &msg[start..end]
+    };
+    let mut payload = {
+        let start = usize::from(
+            expected_message_data_offset
+                .checked_sub(message_offset)
+                .unwrap(),
+        );
+        let end = start
+            .checked_add(expected_message_size.into())
+            .ok_or(SignatureVerificationError::MessageOffsetOverflow)?;
+        &msg[start..end]
+    };
+
+    if public_key != signer {
+        msg!("Invalid Ix: message signed by: {:?}", public_key);
+        msg!("Invalid Ix: expected pubkey: {:?}", signer);
+        return Err(ErrorCode::SigVerificationFailed.into());
     }
+
+    let signature = {
+        let start = usize::from(
+            offsets
+                .signature_offset
+                .checked_sub(message_offset)
+                .unwrap(),
+        );
+        let end = start
+            .checked_add(SIGNATURE_LEN.into())
+            .ok_or(SignatureVerificationError::InvalidSignatureOffset)?;
+        let sig_slice = &msg[start..end];
+        let mut sig_array = [0u8; 64];
+        sig_array.copy_from_slice(sig_slice);
+        sig_array
```

Comment:
nit: same thing without copy
```suggestion
        &msg[start..end].try_into().unwrap()
```